### PR TITLE
feat: Attribute filter/Segment for Round Robin hosts

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

- Adds an option to filter 'All team members' using a segment [Loom Demo](https://www.loom.com/share/3da592fc2efc48f884fce5e655903f7d) - Fixes CAL-4604
- Adds dry-run mode for booking. Allows to be able to test the actual booking without making any DB update and without sending emails. See it in action in above loom.
- A good amount of code movement owing to supporting raqb beyond routing-forms
- Moves 'Value of field' options to the top in select list. Fixes CAL-4567

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
See loom

### CODE REVIEW

File: apps/web/styles/globals.css [hunk 1]
```diff
@@ -483,3 +483,6 @@ select:focus {
 .intercom-lightweight-app  {
   @apply z-40 !important;
 }
+
+
+@import "./raqb.css";
```
Comment (hariombalhara) [inline apps/web/styles/globals.css:488]: We need raqb at more places now. So, it is better to have it in globals.css.

File: packages/app-store/routing-forms/components/react-awesome-query-builder/config/BasicConfig.ts [hunk 2]
```diff
@@ -30,11 +27,19 @@ const conjunctions: Conjunctions = {
     label: "And",
     jsonLogicConj: "and",
     reversedConj: "OR",
+    formatConj: () => "",
+    sqlFormatConj: () => "",
+    spelFormatConj: () => "",
+    mongoConj: "",
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/components/react-awesome-query-builder/config/BasicConfig.ts:33]: raqb methods throw TS error for these. So, let it have them even though we don't need them

File: packages/app-store/routing-forms/components/react-awesome-query-builder/config/BasicConfig.ts [hunk 3]
```diff
@@ -175,7 +180,7 @@ const operators: Operators = {
   },
   select_not_any_in: {
     isNotOp: true,
-    label: "Not in",
+    label: "Not any in",
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/components/react-awesome-query-builder/config/BasicConfig.ts:183]: The reverse operator is already called Any in. So, "Not any in' is correct

File: packages/lib/raqb/findTeamMembersMatchingAttributeLogic.test.ts
Comment (hariombalhara) [inline packages/lib/raqb/findTeamMembersMatchingAttributeLogic.test.ts:1]: Removed Route from the name as it now handles finding matching members for a raqbQueryValue and not route. 

Route was an unnecessary dependency.

File: packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx [hunk 4]
```diff
@@ -748,7 +748,7 @@ const Routes = ({
   const attributesQueryBuilderConfig = attributes
     ? getQueryBuilderConfigForAttributes({
         attributes: attributes,
-        form: hookForm.getValues(),
+        dynamicOperandFields: hookForm.getValues().fields
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx:751]: We just needed fields and that makes `getQueryBuilderConfigForAttributes` independent of Routing Form

File: packages/app-store/routing-forms/trpc/formMutation.handler.ts [hunk 5]
```diff
@@ -24,58 +24,6 @@ interface FormMutationHandlerOptions {
...
-    if (isRouter(route)) {
-      return;
-    }
-    // We use separate schema for queryValye here which is much more strict
-    // It allows that we are still lenient with schema while reading the queryValue but while saving it we are strict
-    const parsedFormFieldsQueryValue = queryValueSaveValidationSchema.safeParse(route.queryValue);
-    if (!parsedFormFieldsQueryValue.success) {
-      throw new TRPCError({
-        code: "BAD_REQUEST",
-        message: `Route ${routeIndex + 1} form fields: ${getErrorMessageFromZodError(
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/trpc/formMutation.handler.ts:45]: Not needed now. As the default zod schema handles the same errors though they might not be that pretty

File: packages/lib/raqb/raqbUtils.ts [hunk 6]
```diff
@@ -237,63 +293,46 @@ function getAttributesData({
...
 
   return attributesQueryValueCompatibleForMatchingWithFormField;
 }
 
-function getAttributesQueryBuilderConfig({
-  form,
+/**
+ * Returns attributesQueryBuilderConfig with the list of labels instead of list of ids.
+ */
+export function getAttributesQueryBuilderConfigHavingListofLabels({
```
Comment (hariombalhara) [inline packages/lib/raqb/raqbUtils.ts:322]: Rename

File: apps/web/components/dialog/RerouteDialog.tsx [hunk 7]
```diff
@@ -775,7 +775,7 @@ const RerouteDialogContentAndFooterWithFormResponse = ({
   >([]);
 
   const findTeamMembersMatchingAttributeLogicMutation =
-    trpc.viewer.appRoutingForms.findTeamMembersMatchingAttributeLogic.useMutation({
+    trpc.viewer.appRoutingForms.findTeamMembersMatchingAttributeLogicOfRoute.useMutation({
```
Comment (hariombalhara) [inline apps/web/components/dialog/RerouteDialog.tsx:778]: The name change reflects that it identifies matching members for the route, establishing a distinction from the similar function findTeamMembersMatchingAttributeLogic, which operates on any attribute query.

File: packages/app-store/routing-forms/trpc/_router.ts [hunk 8]
```diff
@@ -98,12 +98,12 @@ const appRoutingForms = router({
       return handler({ ctx, input });
     }),
 
-  findTeamMembersMatchingAttributeLogic: authedProcedure
+  findTeamMembersMatchingAttributeLogicOfRoute: authedProcedure
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/trpc/_router.ts:101]: As explained earlier it is a rename to convey its distinction from similar trpc route.

File: packages/app-store/routing-forms/trpc/findTeamMembersMatchingAttributeLogicOfRoute.handler.ts [hunk 9]
```diff
@@ -57,11 +57,14 @@ export const findTeamMembersMatchingAttributeLogicHandler = async ({
     checkedFallback,
     mainAttributeLogicBuildingWarnings: mainWarnings,
     fallbackAttributeLogicBuildingWarnings: fallbackWarnings,
-  } = await findTeamMembersMatchingAttributeLogicOfRoute(
+  } = await findTeamMembersMatchingAttributeLogic(
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/trpc/findTeamMembersMatchingAttributeLogicOfRoute.handler.ts:60]: Pass the information in a Routing Form agnostic way now which is also reusable by RR hosts filter

File: packages/features/bookings/lib/handleNewBooking/loadAndValidateUsers.ts [hunk 10]
```diff
@@ -18,7 +18,18 @@ type Users = (Awaited<ReturnType<typeof loadUsers>>[number] & {
...
 
-type EventType = Pick<NewBookingEventType, "hosts" | "users" | "id" | "userId" | "schedulingType">;
+type EventType = Pick<
+  NewBookingEventType,
+  | "hosts"
+  | "users"
+  | "id"
+  | "userId"
+  | "schedulingType"
+  | "team"
```
Comment (hariombalhara) [inline packages/features/bookings/lib/handleNewBooking/loadAndValidateUsers.ts:28]: `team` was already available but just not needed earlier.

File: packages/features/eventtypes/components/AddMembersWithSwitch.tsx [hunk 11]
```diff
@@ -100,61 +105,223 @@ const CheckedHostField = ({
   );
 };
 
-const AddMembersWithSwitch = ({
```
Comment (hariombalhara) [inline packages/features/eventtypes/components/AddMembersWithSwitch.tsx:103]: Moved a bit below

File: packages/features/eventtypes/components/AddMembersWithSwitch.tsx [hunk 12]
```diff
@@ -100,61 +105,223 @@ const CheckedHostField = ({
...
-        ) : (
-          <></>
-        )}
-        {!assignAllTeamMembers || !automaticAddAllEnabled ? (
+  const assignmentState = getAssignmentState({
+    assignAllTeamMembers,
+    assignTeamMembersInSegment,
+    isAssigningAllTeamMembersApplicable: automaticAddAllEnabled,
+    isSegmentApplicable,
+  });
```
Comment (hariombalhara) [inline packages/features/eventtypes/components/AddMembersWithSwitch.tsx:242]: Simplify the UI cases, by creating states.

File: packages/features/eventtypes/components/AddMembersWithSwitch.tsx [hunk 13]
```diff
@@ -100,64 +104,229 @@ const CheckedHostField = ({
...
+        </div>
+        {assignmentState !== AssignmentState.ALL_TEAM_MEMBERS_ENABLED_AND_SEGMENT_NOT_APPLICABLE && (
+          <MembersSegmentWithToggle
+            teamId={teamId}
+            assignRRMembersUsingSegment={assignRRMembersUsingSegment}
+            setassignRRMembersUsingSegment={setassignRRMembersUsingSegment}
+            rrSegmentQueryValue={rrSegmentQueryValue}
+            setrrSegmentQueryValue={setrrSegmentQueryValue}
+          />
         )}
```
Comment (hariombalhara) [inline packages/features/eventtypes/components/AddMembersWithSwitch.tsx:281]: Simplified this by using states. It also makes it easy to add more cases

File: packages/lib/raqb/raqbUtils.ts [hunk 14]
```diff
@@ -1,33 +1,64 @@
...
+  return chosenOptions.map((option) => option.label);
+}
 
 function ensureArray(value: string | string[]) {
   return typeof value === "string" ? [value] : value;
 }
 
-// We connect Form Field value and Attribute value using the labels lowercased
-function compatibleForAttributeAndFormFieldMatch<T extends string | string[]>(
+function caseInsensitive<T extends string | string[]>(
```
Comment (hariombalhara) [inline packages/lib/raqb/raqbUtils.ts:61]: Renamed to make it clear what it does. it only has this job.

**Comment by hariombalhara** [comment]
No. The result is always deterministic. There was a Math.random case which was removed last week. So, if it is not what the dry run says it will be, it is a bug.

**Comment by alishaz-polymath** [comment]
Awesome! Sounds perfect!

**Review by CarinaWolli — COMMENTED** [review]
This doesn't yet work with attribute routing, right? Like, attribute-based routing still expects all members to be assigned to the event type?

Tested attributed filter for assignment, works as expected ✅
Still need to test dry run booking

**Comment by hariombalhara** [comment]
> This doesn't yet work with attribute routing, right? Like, attribute-based routing still expects all members to be assigned to the event type?

Yes, it doesn' t yet.

**Comment by hariombalhara** [comment]
@CarinaWolli @alishaz-polymath Pushed some changes to fix unit tests that were failing across a lot of files. They were just unit tests failures because of how formbricks was being imported in non-react-component tests.

I have fixed it by decoupling RAQB components from the config.

File: packages/app-store/routing-forms/components/FormInputFields.tsx [hunk 15]
```diff
@@ -26,7 +27,10 @@ export type FormInputFieldsProps = {
 export default function FormInputFields(props: FormInputFieldsProps) {
   const { form, response, setResponse, disabledFields = [] } = props;
 
-  const formFieldsQueryBuilderConfig = getQueryBuilderConfigForFormFields(form);
+  const formFieldsQueryBuilderConfig = withRaqbSettingsAndWidgets({
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/components/FormInputFields.tsx:30]: withRaqbSettingsAndWidgets allows us to merge settings and widgets later separately where we need to render the inputs or render raqb rules.

File: packages/app-store/routing-forms/components/react-awesome-query-builder/config/config.ts
Comment (hariombalhara) [inline packages/app-store/routing-forms/components/react-awesome-query-builder/config/config.ts:1]: config.tsx renamed to config.ts and some code moved to uiConfig.tsx

File: packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx [hunk 16]
```diff
@@ -341,20 +346,16 @@ const Route = ({
         }
       : undefined;
 
-  const chosenEventTypeForRedirect = eventTypeRedirectUrlSelectedOption?.eventTypeId
-    ? eventTypesMap.get(eventTypeRedirectUrlSelectedOption.eventTypeId)
-    : null;
-
-  const isRoundRobinEventSelectedForRedirect =
-    chosenEventTypeForRedirect?.schedulingType === SchedulingType.ROUND_ROBIN;
-
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx:350]: Dead code

**Review by CarinaWolli — DISMISSED** [review]
Great Work 👏🏻

File: apps/web/public/static/locales/en/common.json [hunk 17]
```diff
@@ -2682,6 +2682,7 @@
   "booking_limits_updated_successfully": "Booking limits updated successfully",
   "you_are_unauthorized_to_make_this_change_to_the_booking": "You are unauthorized to make this change to the booking",
   "matching_members": "Matching members",
+  "x_matching_members": "{{x}} matching members",
```
Comment (alishaz-polymath) [inline apps/web/public/static/locales/en/common.json:2685]: NIT: probably a good idea to use context to confirm the type of X here for our translation contributors.

**Review by alishaz-polymath — DISMISSED** [review]
Legendary work 🚀🚀🚀

**Review by joeauyeung — APPROVED** [review]

File: packages/features/eventtypes/components/AddMembersWithSwitch.tsx [hunk 18]
```diff
@@ -1,11 +1,15 @@
...
 import type { Options } from "react-select";
 
 import type { FormValues, Host, TeamMember } from "@calcom/features/eventtypes/lib/types";
 import { useLocale } from "@calcom/lib/hooks/useLocale";
-import { Label } from "@calcom/ui";
+import type { AttributesQueryValue } from "@calcom/lib/raqb/types";
+import { trpc } from "@calcom/trpc";
+import { Label, SettingsToggle } from "@calcom/ui";
 
+import { Segment } from "../../../../apps/web/components/Segment";
```
Comment (ThyMinimalDev) [inline packages/features/eventtypes/components/AddMembersWithSwitch.tsx:12]: I believe we should avoid importing apps code into package, always the other way around

Comment (ThyMinimalDev) [reply packages/features/eventtypes/components/AddMembersWithSwitch.tsx:12]: will make a follow up pr to move segment to a package and make it compatible with Platform

**Pull request merged**